### ISSUE TITLE
should meson introspect ignore DESTDIR?

### ISSUE THREAD

**Issue by germandiagogomez** [mesonbuild/meson#6754]
*This is not necessarily a bug*.

Just my two cents here about whether this should be discussed. I am generating a installer for Windows. I need to get the list of installed files to put them in a wix toolset database (this is a xml with all files to be used by the installer).

I want to install the files in DESTDIR as a first step, using `set DESTDIR=%cd%/build/staging` and doing `ninja -C build install` and later gather staged files and put them in a msi installer through a `run_target('create_msi_installer')`. These are the correct files to gather, and not the ones in the build directory actually. But meson introspect ignores DESTDIR.

- Should DESTDIR be ignored by meson introspect (this is what it does, so the installation directory is not correct).

**Comment by mensinda** [comment]
No, `meson introspect` should not handle `DESTDIR`. More generally, the *entire* introspection API should not depend on *any* environment variable. It must only depend on the results of the last meson run. This is because the `meson introspect` command is only a small frontend for the `*.json` files located in the `meson-info` directory. So the only thing `meson introspect --installed <build dir>` does, is to read the correct JSON file (here `<build dir>/meson-info/intro-installed.json`) and paste it to stdout.

The only exception here are those options that can also work without a configured build directory. However, this can be thought of as an entirely different API, since meson does its best to *guess* the correct values and the output format may differ from the normal introspection API.

**Comment by mensinda** [comment]
Also, I might not entirely know what `DESTDIR` does, but can't you just slightly post-process the introspection output by prepending `$DESTDIR/` to each installed file in your installation script?

**Comment by germandiagogomez** [comment]
Hello @mensinda ,

The solution you suggested is what I did. However, handling DESTDIR is not as simple as it would be to invoke meson introspect and has some corner cases (especially on Windows).

I also agree that depending on an environment variable by default would open the door to surprises. However, maybe (and I say *maybe*, since I have no clear idea of the consequences) augmenting in some way `meson introspect --installed`, maybe adding `--no-ignore-destdir` or similar.

The fact here is that staging builds are regularly done and there is where files that will be further packaged live (.deb installer, .msi package or whatever you pack). I think it would be a good idea to have an option to point to those files. This would ease package creation quite a bit.

When I create packages for the different systems (mac, linux, windows, maybe others), I try to take advantage of meson as much as possible and give a last tweak with some script. The less we rely on scripts, the more reliable and portable the solution would be.

**Comment by mensinda** [comment]
After thinking about it, I would say that `meson introspect` definitely should *not* handle `DESTDIR`. Doing so would require duplicating the complete installation logic from `minstall.py` to correctly cover all the corner cases. The only thing I might consider doing (with an opt-in) would be to simply prepend `$DESTDIR` to all the installation paths. However, this is apparently not possible since there are corner cases on Windows (like always).

That doesn't mean that listing all installation files correctly won't be possible, just that the introspection API is not the right place for this. Again a `meson install --dry-run --list` should still be possible.

**Comment by eli-schwartz** [comment]
What are the corner cases, above and beyond using `mesonbuild.scripts.destdir_join(destdir, introspect_path)`?

Not saying that introspect should or shouldn't implement this, but if it did then it shouldn't be *too* hard?
